### DISCUSSION TITLE
Accessing up-to-date atom from within a callback

### DISCUSSION THREAD

**Discussion by LazyAfternoons** [pmndrs/jotai#1390] [Q&A] [answered]
Hello there, this is my first time using a state management for React.
I'm currently writing a React Native application where I'd like to read an atom value from a callback function. 
Currently, I'm using react-usestateref and I'm able to access the up-to-date state value within a callback function. 
Is there a matching pattern I can use with atoms?
Thank you for your help.

**Comment by dai-shi** [comment, marked as answer]
Usually (unless something depends on other libs), we create a write only atom outside components.
```js
const countAtom = atom(0)

const doSomethingAtom = atom(null, (get, set, arg) => {
  const count = get(countAtom) // <--- this is up to date
  console.log(count)
  // ...
})

const Component = () => {
  const doSomething = useSetAtom(doSomethingAtom)
  const handleClick = () => {
    // ...
    doSomething()
  }
  return (
    ...
      <Button onClick={handleClick}>...</Button>
    ...
  )
}
```

**Reply by LazyAfternoons** [reply]
Thank you for your help and for the awesome work on Jotai. 
I can't really fit your example, probably because I don't understand it properly. I thought I would provide a very simplified example of how I'm achiving it with react-usestateref. 
Basically I fetch the initial state from a database and then I subscribe to changes providing a callback. 
This callback (handleUpdate) fetches the delta and updates the 'requests' state:

```
const Requests = () => {
  const [requests, setRequests, requestsRef] = useState([]);

  const handleUpdate = async (payload) => {
    const req = await getRemoteRequestsUpdate();
    const updatedReqs = [...requestsRef.current, req];
    setRequests(updatedReqs);
  };

  const fetchRequests = async () => {
    try {
      var reqs = await getRemoteRequests();
      setRequests(reqs);
    } catch (error) {
      console.log(error);
    }
  };

  useEffect(() => {
    fetchRequests();
    const requestsChannel = requestsUpdateChannel(handleUpdate);
    
    return () => {
      removeChannel(requestsChannel);
    };
  }, []);

  return ( 
    ...
  );
};
```

Thank you again!

**Reply by dai-shi** [reply]
That specific case can be done with function update:
```js
  const handleUpdate = async (payload) => {
    const req = await getRemoteRequestsUpdate();
    setRequests(prev => [...prev, req]);
  };
```

**Reply by LazyAfternoons** [reply]
Thank you, works perfectly.
